---
title: "Bioinformatics Project II"
author: "Emily Mynar & Anna Rees"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE,
                      # warning = FALSE,
                      # message = FALSE,
                      fig.align = "center",
                      fig.height = 8)

pacman::p_load(dplyr, ggplot2, DESeq2, tidyverse, forcats, readr, usedist)

theme_set(theme_bw())
```

## 


```{r read files}

meta <- read.table(file = './Data/SRP074247/metadata_SRP074247.tsv',
           sep = '\t',
           header = TRUE)

janitor::clean_names(meta)
```



```{r clean meta file}
meta_clean <-
  
meta |>
  # Changing columns to choose based on how our dataset is stored
  mutate('sample' = refinebio_accession_code,
         'title' = refinebio_title,
         'treatment' = refinebio_title,
         'replicate' = refinebio_title,
         )

for(i in 1:nrow(meta_clean)){
  if(grepl('Mock', meta_clean$title[i])){
    meta_clean$treatment[i] <- 'Mock'}
  if(grepl('with Reovirus', meta_clean$title[i])){
    meta_clean$treatment[i] <- 'Reovirus'}
  if(grepl('Mutant', meta_clean$title[i])){
    meta_clean$treatment[i] <- 'Mutant'}
}

for(i in 1:nrow(meta_clean)){
  if(grepl('1', meta_clean$title[i])){
    meta_clean$replicate[i] <- '1'}
  if(grepl('2', meta_clean$title[i])){
    meta_clean$replicate[i] <- '2'}
  if(grepl('3', meta_clean$title[i])){
    meta_clean$replicate[i] <- '3'}
}

meta_clean <-
  meta_clean |>
  select(sample, title, treatment, replicate) |>
  janitor::clean_names()

meta_clean
```
```{r read counts matrix}
# Read in counts matrix
data <- read.table(file = './Data/SRP074247/SRP074247.tsv',
           sep = '\t',
           header = TRUE,
           row.names = 1)

tibble(data)
```
```{r check identical}

identical(colnames(data), meta_clean$sample)

```

```{r change column names}

colnames(data) <- paste(meta_clean$treatment, meta_clean$replicate, sep = " ")

tibble(data)
```

```{r add gene as column}

data <- rownames_to_column(data, 'gene')

tibble(data)
```

```{r pivot data}

data_long <-
  pivot_longer(data = data,
               cols = -gene,
               names_to = 'ID',
               values_to = 'reads') |>
  
  mutate('sample' = ID) |>

  separate(col = 'ID',
           into = c('treatment', 'replicate'),
           sep = ' ') |>
  
  mutate(treatment = as.factor(treatment),
         replicate = as.factor(replicate)) |>
  
  select(sample, treatment, replicate, gene, reads)

tibble(data_long)

```

```{r colors}
colors <- c('Mock' = '#FA8633',
            'Mutant' = '#1C6C7F',
            'Reovirus' = '#9A6DA3')
```

```{r check quality}

sum_by_sample <-
  data_long |> 
  group_by(treatment, sample) |> 
  summarize('sum' = sum(reads)) |>
  ungroup()

ggplot(data = data_long,
       mapping = aes(x = sample)) +
          
  geom_col(data = sum_by_sample,
           mapping = aes(y = sum,
                         fill = treatment),
           color = 'black') +
  
  labs(x = 'Sample',
       y = 'Total Reads',
       title = 'Assessing the quality of the dataset using number of reads',
       caption = 'Data from refine.bio\nAccession ID: SRP074247',
       fill = 'Sample type') +
  
  theme(plot.title = element_text(hjust = 0.5,
                                  size = 16),
        axis.text.x = element_text(angle = 90, 
                                   vjust = 0.5, 
                                   hjust = 1),
        axis.title.y = element_text(size = 14),
        axis.title.x = element_text(size = 14),
        legend.position = 'right',
        plot.caption = element_text(face = 'italic')) +
  
  scale_fill_manual(values = colors) +
  
  scale_y_continuous(expand = c(0.01, 0))
```

```{r boxplot}

ggplot(data = sum_by_sample |> group_by(treatment) |> summarize('sum' = sum(sum)),
       mapping = aes(x = sum, 
                     y = treatment,
                     fill = treatment)) +
  
  geom_boxplot(data = data_long |> group_by(reads),
               mapping = aes(x = log2(reads))) +
  
  
  scale_fill_manual(values = colors) +
  
  labs(title = 'Number of reads by sample type',
       x = 'log2(reads)',
       y = NULL,
       fill = 'Sample type') +
  
  theme(plot.title = element_text(hjust = 0.5,
                                  size = 16))
         
         
        
```

```{r t test}

# TODO YOU CANT DO A T TEST WITH MORE THAN TWO LEVELS SO WE WILL COME BACK TO THIS 
```

```{r low abundance histogram}

sum_by_gene <-
  data_long |> 
  group_by(gene) |> 
  summarize('sum' = sum(reads)) |>
  ungroup()

ggplot(data = sum_by_gene,
       mapping = aes(x = log2(sum))) +
  
  geom_histogram(bins = 15,
                 color = 'black',
                 fill = '#1C6C7F') +
  
  labs(x = 'Log-2 transformed counts per gene',
       y = 'Number of genes',
       title = 'Number of reads by gene') +
  
   theme(plot.title = element_text(hjust = 0.5,
                                  size = 16))

```
```{r exclude low abundance}

#high_abundance_genes <-
  #data |>
  #filter_if(is.numeric, all_vars((.) != 0))
  #filter(if_any(-gene, ~ !.x  %in% 0)) # why isn't it getting rid of some of the rows with zero

high_abundance_genes <-
  data_long |>
  pivot_wider(names_from = gene,
              values_from = reads) |>
  select(where(~ all(. != 0)))

high_abundance_long <- 
  high_abundance_genes |>
  pivot_longer(cols = -c(sample, treatment, replicate),
               names_to = 'gene',
               values_to = 'reads')

high_abundance_sum <-
  high_abundance_long |>
  group_by(gene) |> 
  summarize('sum' = sum(reads)) |>
  ungroup()

ggplot(data = high_abundance_sum,
       mapping = aes(x = log2(sum))) +
  
  geom_histogram(bins = 15,
                 color = 'black',
                 fill = '#1C6C7F') +
  
  labs(x = 'Log-2 transformed counts per gene',
       y = 'Number of genes',
       title = 'Number of reads by gene (excluding low abundance genes)') +
  
   theme(plot.title = element_text(hjust = 0.5,
                                  size = 16))


```
```{r distribution per sample}

ggplot(data = high_abundance_long,
       mapping = aes(x = sample)) +
          
  geom_boxplot(mapping = aes(y = log2(reads),
                             fill = treatment),
              color = 'black') +
  
  labs(x = 'Sample ID',
       y = 'Log 2 transformed counts per gene',
       title = 'Number of reads per sample',
       caption = 'Data from refine.bio\nAccession ID: SRP074247',
       fill = 'Sample type') +
  
  theme(plot.title = element_text(hjust = 0.5,
                                  size = 16),
        axis.text.x = element_text(angle = 90, 
                                   vjust = 0.5, 
                                   hjust = 1),
        axis.title.y = element_text(size = 14),
        axis.title.x = element_text(size = 14),
        legend.position = 'right',
        plot.caption = element_text(face = 'italic')) +
  
  scale_fill_manual(values = colors) +
  
  scale_y_continuous(expand = c(0.01, 0))
```

```{r}
sample_names <- c('Mock 1', 'Mock 2', 'Mock 3', 'Mutant 1', 'Mutant 2', 'Mutant 3', 'Reovirus 1', 'Reovirus 2', 'Reovirus 3')
```


```{r sample comparison}
high_abundance_log2 <-
  high_abundance_long |> 
  mutate(reads = log2(reads)) |>
  pivot_wider(names_from = gene,
              values_from = reads) |>
  select(-c(treatment, replicate))

data_dist <- dist(high_abundance_log2, method = "euclidean")
data_dist <- dist_setNames(data_dist, c('Mock 1', 'Mock 2', 'Mock 3', 'Mutant 1', 'Mutant 2', 'Mutant 3', 'Reovirus 1', 'Reovirus 2', 'Reovirus 3'))


plot(hclust(data_dist, method = "complete"), xlab = "Sample", main = NULL)
```

```{r normalizing}

sample_medians <- as.numeric(apply(t(high_abundance_log2), 2, median))

grand_median <- mean(sample_medians)

correction_factors <- grand_median - sample_medians

normalizing <- data.frame(sample_names, correction_factors)

normalizing
```

```{r}

```




words at the end!
